src/twisted/test/test_iosim.py

```diff
@@ -39,3 +44,85 @@ def test_writeSequence(self):
         a.writeSequence([b"b", b"c", b"d"])
 
         self.assertEqual(b"".join(a.stream), b"abcd")
+
+
+
+@implementer(IPushProducer)
+class StrictPushProducer(object):
+    """
+    A L{IPushProducer} implementation which produces nothing but enforces
+    preconditions on its state transition methods.
+    """
+    _state = u"running"
+
+    def stopProducing(self):
+        if self._state == u"stopped":
+            raise ValueError(u"Cannot stop already-stopped IPushProducer")
+        self._state = u"stopped"
+
+
+    def pauseProducing(self):
+        if self._state != u"running":
+            raise ValueError(
+                u"Cannot pause {} IPushProducer".format(self._state)
+            )
+        self._state = u"paused"
+
+
+    def resumeProducing(self):
+        if self._state != u"paused":
+            raise ValueError(
+                u"Cannot resume {} IPushProducer".format(self._state)
+            )
+        self._state = u"running"
+
+
+
+class IOPumpTests(TestCase):
+    """
+    Tests for L{IOPump}.
+    """
+    def _testStreamingProducer(self, mode):
```

Comment:
I'm going to guess that twistedchecker yells about this being undocumented, unless we disabled it again.